mono/metadata/native-library.c

```diff
@@ -408,138 +433,447 @@ mono_lookup_pinvoke_call_internal (MonoMethod *method, MonoError *error)
+	ERROR_DECL (error);
+
+	result = netcore_resolve_with_dll_import_resolver (alc, assembly, scope, flags, error);
+	if (!is_ok (error))
+		mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_ASSEMBLY, "Error while invoking ALC DllImportResolver(\"%s\") delegate: '%s'", scope, mono_error_get_message (error));
+
+	mono_error_cleanup (error);
+
+	return result;
+}
+
+static MonoDl *
+netcore_resolve_with_load (MonoAssemblyLoadContext *alc, const char *scope, MonoError *error)
+{
+	MonoDl *result = NULL;
+	gpointer lib = NULL;
+
+	static MonoMethod *resolve;
+
+	if (!resolve) {
+		ERROR_DECL (local_error);
+		MonoClass *alc_class = mono_class_get_assembly_load_context_class ();
+		g_assert (alc_class);
+		MonoMethod *m = mono_class_get_method_from_name_checked (alc_class, "MonoResolveUnmanagedDll", -1, 0, local_error);
+		mono_error_assert_ok (local_error);
+		resolve = m;
 	}
+	g_assert (resolve);
 
-	if (!cached) {
-		mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT,
-					"DllImport loaded library '%s'.", found_name);
-		mono_image_lock (image);
-		if (!g_hash_table_lookup (image->pinvoke_scopes, new_scope)) {
-			g_hash_table_insert (image->pinvoke_scopes, g_strdup (new_scope), module);
-			g_hash_table_insert (image->pinvoke_scope_filenames, g_strdup (new_scope), g_strdup (found_name));
-		}
-		mono_image_unlock (image);
+	if (mono_runtime_get_no_exec ())
+		return NULL;
+
+	HANDLE_FUNCTION_ENTER ();
+
+	MonoStringHandle scope_handle = mono_string_new_handle (mono_alc_domain (alc), scope, error);
+	goto_if_nok (error, leave);
+
+	gpointer gchandle = GUINT_TO_POINTER (alc->gchandle);
+	gpointer args [3];
+	args [0] = &scope_handle;
+	args [1] = &gchandle;
+	args [2] = &lib;
+	mono_runtime_invoke_checked (resolve, NULL, args, error);
+	goto_if_nok (error, leave);
+
+	native_library_lock ();
+	result = netcore_handle_lookup (lib);
+	native_library_unlock ();
+
+leave:
+	HANDLE_FUNCTION_RETURN_VAL (result);
+}
+
+static MonoDl *
+netcore_resolve_with_load_nofail (MonoAssemblyLoadContext *alc, const char *scope)
+{
+	MonoDl *result = NULL;
+	ERROR_DECL (error);
+
+	result = netcore_resolve_with_load (alc, scope, error);
+	if (!is_ok (error))
+		mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_ASSEMBLY, "Error while invoking ALC LoadUnmanagedDll(\"%s\") method: '%s'", scope, mono_error_get_message (error));
+
+	mono_error_cleanup (error);
+
+	return result;
+}
+
+static MonoDl *
+netcore_resolve_with_resolving_event (MonoAssemblyLoadContext *alc, MonoAssembly *assembly, const char *scope, MonoError *error)
+{
+	MonoDl *result = NULL;
+	gpointer lib = NULL;
+	MonoDomain *domain = mono_alc_domain (alc);
+
+	static MonoMethod *resolve;
+
+	if (!resolve) {
+		ERROR_DECL (local_error);
+		MonoClass *alc_class = mono_class_get_assembly_load_context_class ();
+		g_assert (alc_class);
+		MonoMethod *m = mono_class_get_method_from_name_checked (alc_class, "MonoResolveUnmanagedDllUsingEvent", -1, 0, local_error);
+		mono_error_assert_ok (local_error);
+		resolve = m;
```

Comment:
Missing barriers.

Reply:
Why do I need barriers here?

Reply:
These need barriers because the cpu might reorder stores such that the MonoMethod structure pointed to by 'm' is not fully initialized, and it becomes visible to other threads because its assigned to the global 'resolve' variable.

Reply:
Oh right, that's static.
This is... a bit of a problem, since that particular pattern (without barriers) is used elsewhere as well. I guess I'll have to patch it up everywhere.

Reply:
To be clear, elsewhere meaning code prior to this PR.

Reply:
Its not very likely to be a problem but better be safe than sorry.

Reply:
Agreed. I'll fix for this PR now and put fixing across the codebase on my backlog.

Reply:
OK, I lied and I'll fix this with the rest of them, I want to get this merged.